chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,350 @@
+
+        # Expand prediction and broadcast instances to 3rd axis
+        p = F.reshape(pred, (bs, n_filters, 1, n_loc))
+        p = F.broadcast_to(p, (bs, n_filters, n_instances, n_loc))
+
+        # Expand ground truth to match the size but do not broadcast
+        g = F.reshape(gt, (bs, 1, n_instances, n_loc))
+
+        p = p * F.cast(g, p.dtype)
+        module = cuda.get_array_module(p)
+
+        means = []
+        for i in range(bs):
+
+            p_item = p[i, :, gt_idx[i]]
+            g_item = g[i, :, gt_idx[i]]
+
+            p_sum = F.sum(p_item, axis=2)
+            g_sum = F.cast(F.sum(g_item, axis=2), p_sum.dtype)
+            _mean_sample = p_sum / g_sum
+
+            n_fill_objects = max_n_objects - n_objects[i]
+
+            if n_fill_objects != 0:
+                _fill_sample = module.zeros(
+                    (module.asnumpy(n_fill_objects), n_filters),
+                    dtype=_mean_sample.dtype)
+                _mean_sample = F.concat((_mean_sample, _fill_sample), axis=0)
+
+            means.append(_mean_sample)
+
+        means = F.stack(means)
+        return means
+
+    def _prepare_inputs(self, prediction, labels):
+        """Function to preprocess inputs
+
+        Args:
+            prediction (nd-array): Prediction output
+            labels (nd-array): Ground truth output
+        Returns:
+            nd-array : Prediction output
+            nd-array : Ground truth output
+        """
+        # Reshape layers to prepare for processing
+
+        p_shape = prediction.shape
+        l_shape = labels.shape
+
+        prediction = F.reshape(prediction, shape=(
+            p_shape[0], p_shape[1], p_shape[2] * p_shape[3]))
+        labels = F.reshape(labels, shape=(
+            l_shape[0], l_shape[1], l_shape[2] * l_shape[3]))
+
+        return prediction, labels
+
+    def apply(self, *x):
+        """Initial apply function
+
+        Args:
+            x (tuple) : Contains several inputs
+                - x[0] = segmentation prediction output
+                - x[1] = segmentation ground truth
+                - x[2] = number of objects in ground truth
+                - x[3] = indexes of non-zero ground truths
+        Returns:
+            float : Loss value
+        """
+
+        # Inputs
+        prediction, labels = self._prepare_inputs(x[0], x[1])
+        n_objects, gt_idx = x[2:4]
+
+        # Calculate cluster means
+        c_means = self._means(prediction, labels, n_objects,
+                              self.max_n_clusters, gt_idx)
+
+        # Calculate losses
+        l_var = self._variance_term(prediction, labels, c_means, self.delta_v,
+                                    gt_idx)
+        l_dist = self._distance_term(c_means, self.delta_d, n_objects)
+        l_reg = self._regularization_term(c_means, n_objects)
+
+        return self.alpha * l_var + self.beta * l_dist + self.gamma * l_reg
+
+    def __call__(self, *args):
+        return self.apply(*args)
```

Comment:
Could you let me know why you prepare `apply` method? If the reason is to align with implementation of `FunctionNode`, I think we do not have to do that because this class is not a subclass of `FunctionNode`. Just providing `__call__` method would be enough.